core/src/main/scala/scalaz/Cord.scala

```diff
@@ -1,163 +1,140 @@
-   */
-  def size: Int = self.measure
-
-  /**
-   * Appends another `Cord` to the end of this one.
-   * Time complexity: O(log (min N M)) where M and N are the lengths of the two `Cord`s.
-   */
-  def ++(xs: Cord): Cord = cord(self <++> xs.self)
-
-  /**
-   * Appends a `String` to the end of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def :+(x: => String): Cord = cord(self :+ x)
-
-  /**
-   * Prepends a `String` to the beginning of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def +:(x: String): Cord = cord(x +: self)
-
-  /**
-   * Prepends a `Char` to the beginning of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def -:(x: Char): Cord = cord(String.valueOf(x) +: self)
-
-  /**
-   * Appends a `Char` to the end of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def :-(x: => Char): Cord = cord(self :+ String.valueOf(x))
-
-  /**
-   * Removes the first character of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def tail: Cord = drop(1)
-
-  /**
-   * Removes the last character of this `Cord`.
-   * Time complexity: O(1)
-   */
-  def init: Cord = take(self.measure - 1)
-
-  /**
-   * Removes the first `n` characters from the front of this `Cord`.
-   * Time complexity: O(min N (N - n))
-   */
-  def drop(n: Int): Cord = split(n)._2
-
-  /**
-   * Returns the first `n` characters at the front of this `Cord`.
-   * Time complexity: O(min N (N - n))
-   */
-  def take(n: Int): Cord = split(n)._1
-
-  /**
-   * Modifies each character in this `Cord` by the given function.
-   * Time complexity: O(N)
-   */
-  def map(f: Char => Char): Cord = cord(self map (_ map f))
-
-  def toList: List[Char] = toVector.toList
-  def toStream: Stream[Char] = toVector.toStream
-  def toVector: Vector[Char] = self.foldMap(_.toVector)
-  override def toString: String = {
-    val sb = new java.lang.StringBuilder(self.measure)
-    self.foreach { x =>
-      val _ = sb append x
-    }
-    sb.toString
-  }
-
-  /** Transforms each character to a `Cord` according to the given function and concatenates them all into one `Cord`. */
-  def flatMap(f: Char => Cord): Cord = toVector.foldLeft(Cord())((as, a) => as ++ f(a))
-
-  /** Returns whether this `Cord` will expand to an empty string. */
-  def isEmpty: Boolean = !nonEmpty
-
-  /** Returns whether this `Cord` will expand to a non-empty string. */
-  def nonEmpty: Boolean = self.iterator.exists(_.nonEmpty)
+  /** Strict evaluation variant of `Monoid.append`. */
+  final def ++(o: Cord): Cord = Cord.Branch(this, o)
 }
-
 object Cord {
-  private def cord[A](v: FingerTree[Int, String]): Cord = new Cord(v)
-
-  implicit def stringToCord(s: String): Cord = cord(FingerTree.single[Int, String](s))
-
-  lazy val empty: Cord = apply()
-
-  def apply(as: Cord*): Cord = as.foldLeft(cord(FingerTree.empty))(_ ++ _)
+  def apply(s: String): Cord = Leaf.apply(s)
+  def apply(): Cord = Leaf.Empty
+
+  private[scalaz] final case class Leaf private (
+    s: String
+  ) extends Cord
```

Comment:
perhaps this could be a AnyVal ?

Reply:
Would `AnyVal` work properly here? Honest question: I know it's flakey and I can't remember when it detriggers.

Reply:
I think its mostly when in collections and sometimes passed to functions that it gets boxed. Im pretty sure when referenced in a case class like this it saves a pointer. I think this is the ideal case.
Im not sure on all the caveats, but I'd think it'd do more good than harm.

Reply:
I don't understand the reason for AnyVal. Is that to include legacy toString methods? I don't think it's worth it for the loss of the type system.

Reply:
Oh you mean for the product. No you can't do that on adt members. Maybe in dotty with union types, who knows.

Reply:
Oh.. thanks fommil. I thought it was possible. Never mind.

Reply:
Although my misunderstanding does raise an interesting optimisation: DelayedLeaf that takes an Any and calls toString on it during shows. This could break the Monoid laws (no way to tell if empty) unless we say that an empty string is not an empty Cord. Which is funky.